icu4j/main/classes/core/src/com/ibm/icu/impl/personname/PersonNameFormatterImpl.java

```diff
@@ -268,6 +279,30 @@ private PersonNamePattern getBestPattern(PersonNamePattern[] patterns, PersonNam
         }
     }
 
+    /**
+     * Internal function to figure out the name's script by examining its characters.
+     * @param name The name for which we need the script
+     * @return The four-letter script code for the name.
+     */
+    private String getNameScript(PersonName name) {
+        // Rather than exhaustively checking all the fields in the name, we just check the given-name
+        // and surname fields, giving preference to the script of the surname if they're different
+        // (we concatenate them into one string for simplicity).  The "name script" is the script
+        // of the first character we find whose script isn't "common".  If that script is one
+        // of the scripts used by the specified locale, we have a match.
+        String nameText = name.getFieldValue(PersonName.NameField.GIVEN, Collections.emptySet())
+                + name.getFieldValue(PersonName.NameField.SURNAME, Collections.emptySet());
+        int stringScript = UScript.UNKNOWN;
+        for (int i = 0; stringScript == UScript.UNKNOWN && i < nameText.length(); i++) {
+            char c = nameText.charAt(i);
+            int charScript = UScript.getScript(c);
+            if (charScript != UScript.COMMON && charScript != UScript.INHERITED) {
+                stringScript = charScript;
```

Comment:
This stops at the first "strong" character. Probably ok.
Might be a bit better to keep going just in case we hit a character whose script is compatible with the name locale or with the formatter locale. But we can discuss that later.

Reply:
At today's Person Names meeting, we agreed to punt on this request for now, just for performance and simplicity, on the ground that mixed script name input should be really rare in practice.  But we all acknowledged that this might come back in a future release if we turn out to be wrong about this.

We also discussed putting the surname before the given name and all agreed that made sense.  Amusingly, the doc comment at the top of the function already says we're doing that ("...we just check the given-name and surname fields, _giving preference to the script of the surname if they're different_..."), so my code didn't match my documentation.  I fixed the code.